### DISCUSSION TITLE
Non blocking motor movement

### DISCUSSION THREAD

**Discussion by ShakyMUC** [gin66/FastAccelStepper#161] [Q&A] [answered]
Hello everyone,
I used the original AccelStepper Lib on a previous project. Now I need higher speeds and serial communication while the stepper is running.
I tested the maximum speed with FastAccelStepper and one of the examples - all perfect.
But I have problems with the serial connection of a Nextion display. The serial interface is only polled at the end points of motor movement.
The description says:
"speed/acceleration can be varied while stepper is running (call to functions move or moveTo is needed in order to apply the new values)"

Below my test code. Can anyone tell/help me how to use the non-blocking function of the lib to poll the serial port while the motor is moving and change the motor speed?

Thanks a lot

`
#include <Arduino.h>
#include <FastAccelStepper.h>

#define dirPinStepper    4
#define enablePinStepper 0
#define stepPinStepper   2

FastAccelStepperEngine engine = FastAccelStepperEngine();
FastAccelStepper *stepper = NULL;

// --------- Eigene Definitionen ----------

String buttons_read = ""; // Zwischenspreeicher Auslesen Button
int motorspeed = 10; // Motordrehzahl in %
int gaugespeed = 10; // Anzeigedrehzahl = motorspeed+1.8 für Zeiger auf NEXTION

void setup() {

    engine.init();
    stepper = engine.stepperConnectToPin(stepPinStepper);
    if (stepper) {
       stepper->setDirectionPin(dirPinStepper);
       stepper->setEnablePin(enablePinStepper);
       stepper->setAutoEnable(true);

       stepper->setSpeedInHz(1500);
       stepper->setAcceleration(250000);
    }
       
    Serial.begin(115200);
    Serial2.begin(115200);
    delay(20);
    
    Serial2.print("progressSpeed.val=");
    gaugespeed = motorspeed*1.8;
    Serial2.print(gaugespeed);
    Serial2.write(0xFF);
    Serial2.write(0xFF);
    Serial2.write(0xFF);
    delay(5);

}

void serialRunWatch() {
    buttons_read += char (Serial2.read());
    if (buttons_read == "D")
    {
       motorspeed = motorspeed - 10;
       Serial2.flush();
    }

    if (buttons_read == "U")
    {
      motorspeed = motorspeed + 10;
      Serial2.flush();
    }   
    
    Serial.print("Motorspeed: ");
    Serial.println(motorspeed);
    Serial2.print("progressSpeed.val=");
    gaugespeed = motorspeed*1.8;
    Serial2.print(gaugespeed);
    Serial2.write(0xFF);
    Serial2.write(0xFF);
    Serial2.write(0xFF);
    delay(5); 

    buttons_read = "";  
}

 void loop() {
    stepper->moveTo(2550, true);
    stepper->moveTo(0, true);

    Serial.println("Testmessage");
    serialRunWatch();
 }
`

**Comment by gin66** [comment]
Apparently the bullet list in the readme is not complete. In the API explanation is written for acceleration: „ New value will be used after call to move/moveTo/runForward/runBackward/applySpeedAcceleration/moveByAcceleration“. Especially applySpeedAcceleration() is most straightforward.

If using non-blocking, then two functions come into play to get the status of the motor: isRunning() and getCurrentPosition(). If you move the motor without ˋtrueˋ as second parameter, then the program will just continue to run. If you like to queue commands from the serial, then you would need to wait using isRunning() before calling the next move(). If you like the motor execute immediately commands from serial, then just call move() disregarding the current running status of the motor. FastAccelStepper will perform the needed acceleration/deceleration magic/math for you.

**Comment by ShakyMUC** [comment, marked as answer]
Hello gin,
and first of all: Thank you for the library itself and the support 👍

Over night (a lot of coffee - little sleep) I found out a lot more. Also the variable true/false within moveTo() to turn blocking on or off.

In the meantime I've written some code around it and have the motor control (quite well) under control.
I'm not fundamentally & totally new to programming. BUT - it's just been 30 years since my last times.

Now I'm going to develop everything further so that I have a solid basis for controlling the stepper with dynamic changes (via serial control by the nextion display) of speed/target points while running.
Hopefully I can do that, because I don't want to bore you with stupid "beginner's questions".

After that it would be great if I could post the relevant parts of the code here again, so that you or everyone else can take a look and give me tips on where/how I can optimize something.
